authorization/token/service/token_service.go

```diff
@@ -0,0 +1,226 @@
+			// return a WWW-Authenticate: LOGIN response
+			// TODO adjust this error return code
+			return "", errors.NewInternalError(ctx, nil)
+		}
+
+		// If the token is stale, yet the resource exists in the token
+		// we can re-evaluate its privileges to determine whether they have changed.
+		// If the privileges are unchanged, then reset the token status
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_STALE) && resourceExistsInToken {
+			// Query for all of the token's privileges
+			privileges, err := s.Repositories().TokenRepository().ListPrivileges(ctx, tokenID)
+			if err != nil {
+				return "", errors.NewInternalError(ctx, err)
+			}
+
+			// First we recalculate any stale privileges
+			for _, priv := range privileges {
+				if priv.Stale {
+					// Retrieve the up to date scopes for the resource
+					scopes, err := s.Services().PrivilegeCacheService().ScopesForResource(ctx, priv.IdentityID, priv.ResourceID)
+					if err != nil {
+						return "", errors.NewInternalError(ctx, err)
+					}
+
+					scopesChanged := false
+
+					// Compare the scopes to those contained in the current token
+					for _, tokenPermission := range *tokenClaims.Permissions {
+						// Find the corresponding resource set ID in the token's permissions claim
+						if *tokenPermission.ResourceSetID == priv.ResourceID {
+							if !s.scopesEquivalent(tokenPermission.Scopes, scopes) {
+								scopesChanged = true
+								break
+							}
+						}
+					}
+
+					// If the scopes haven't changed, and the specified resouce ID is already contained in the current
+					// token, then reset the token status to valid and return an empty string
+					if !scopesChanged {
+						loadedToken.Status = 0
+						err = s.Repositories().TokenRepository().Save(ctx, loadedToken)
+						if err != nil {
+							return "", errors.NewInternalError(ctx, err)
+						}
+
+						return "", nil
+					}
+				}
+			}
+		}
+	}
+
+	// If we've gotten this far, it means that either no existing token was found, or the token that was found
+	// has been marked with status STALE and its privileges have changed, in either case we must generate a new token
+	perms := []token.Permissions{}
+
+	// Populate the permissions
+	// TODO populate permissions array
+
+	generatedToken, err := manager.GenerateUnsignedRPTTokenForIdentity(ctx, tokenClaims, *identity, &perms)
+	if err != nil {
+		return "", errors.NewInternalError(ctx, err)
+	}
+
+	// We need to extract the new jti claim from the token
+	claims := generatedToken.Claims.(token.TokenClaims)
+
+	newTokenID, err := uuid.FromString(claims.Id)
+	if err != nil {
+		return "", errors.NewInternalError(ctx, err)
+	}
+
+	// Create a new Token record in the database
+	newTokenRecord := &tokenRepo.Token{
+		TokenID:   newTokenID,
+		Status:    0,
+		TokenType: tokenPkg.TOKEN_TYPE_RPT,
+		// TODO calculate the expiry time
+		/*ExpiryTime: claims.ExpiresAt,*/
+	}
+
+	s.Repositories().TokenRepository().Create(ctx, newTokenRecord)
```

Comment:
Error return value of `(github.com/fabric8-services/fabric8-auth/authorization/token/repository.TokenRepository).Create` is not checked